### DISCUSSION TITLE
Can Deta Micros install Puppeteer?

### DISCUSSION THREAD

**Discussion by azickri** [deta/cloud-docs#538] [Help]
Hello, i want to ask about Deta Micros.
Can i install Puppeteer in deta micros? Because, when i try to install that package and deta deploy that error for unknow reason.

**Comment by stefankuehnel** [comment]
# Deprecated

The following response contains deprecated dependencies. Furthermore, the code is not compatible with current versions of `puppeteer-core`.

Please refer to the following (updated) repository https://github.com/StefanZoneExamples/deta-puppeteer.

---

Yes it is possible, although the solution is a bit strange.

The solution is to use [`puppeteer-core`](https://npm.im/puppeteer-core) with [`chrome-aws-lambda`](https://npm.im/chrome-aws-lambda). Check out [this blog post](https://bitsofco.de/how-to-use-puppeteer-in-a-netlify-aws-lambda-function/) for further details.

### Steps

First install the required packages. I'm assuming you are using [`express`](https://docs.deta.sh/docs/micros/getting_started#creating-your-first-micro) in your Deta Micro.

```
npm i -D puppeteer
npm i puppeteer-core chrome-aws-lambda
```

Next,  in your `index.js`, access Puppeteer via the `chromium` package.

> **Note**: This will determine which Puppeteer package to use. Therefore you can run puppeteer both locally and in production without additional effort.

```js
const express = require("express");
const chromium = require('chrome-aws-lambda');

const app = express();

app.get("/screenshot", async (request, response) => {
	const executablePath = await chromium.executablePath;

	const browser = await chromium.puppeteer.launch({
		args: chromium.args,
		executablePath: executablePath,
		headless: chromium.headless,
	});

	try {
		const page = await browser.newPage();

		await page.goto("https://example.com");

		const screenshot = await page.screenshot({ fullPage: true });

		// IMPORTANT: https://docs.deta.sh/docs/common_issues/#nodejs-micros-cannot-serve-binary-files
		response.type("png");
		response.send(screenshot);
	} catch (error) {
		res.send("An error occured.");
	} finally {
		await browser.close();
	}
});

const port = 3000;

if (!process.env.DETA_RUNTIME) {
    app.listen(port, () => console.log("Started on http://localhost:3000"));
}

module.exports = app;
```

Finally, it is **VERY** important to note that Deta Micros [cannot handle binary files](https://docs.deta.sh/docs/common_issues/#nodejs-micros-cannot-serve-binary-files). I spent some time banging my head on the table trying to solve this problem. But it is easily solvable.

First create a `.env` file in the project root directory.

```
BINARY_CONTENT_TYPES=image/*
```

> **Note**: This enables your Deta Micro to handle files of the `image/*` type, i. e. `image/png`.

Then, run the following command with the Deta CLI.

```sh
deta update -e .env
```

And then it should work. If not, feel free to contact me. I'm running Puppeteer successfully on a Deta Micro. 😃

**Reply by suntong** [reply]
What's the access url?

When I tried it, by visiting my duplicated  a..0.deta.dev, I'm getting a "null" in json; if adding ":3000", I'm getting:

The connection has timed out

An error occurred during a connection to a..0.deta.dev:3000

Any way to troubleshoot it pls?

```
$ deta details
{
        "name": "xx",
        "id": "c9xxx1d",
        "project": "default",
        "runtime": "nodejs14.x",
        "endpoint": "https://axxx0.deta.dev",
        "region": "us-west-1",
        "dependencies": [
                "chrome-aws-lambda@^10.1.0",
                "puppeteer-core@^10.4.0",
                "express@^4.18.2"
        ],
        "environment_variables": [
                "BINARY_CONTENT_TYPES"
        ],
        "visor": "disabled",
        "http_auth": "disabled"
}
```

**Comment by stefankuehnel** [comment]
Hey there, specifying the port `3000` is only necessary if you run the project locally. It does not work in production. 

To troubleshoot, please try calling your endpoint with the `/screenshot` path, i. e. `https://axxx0.deta.dev/screenshot`.

**Reply by stefankuehnel** [reply]
I've looked into the issue and I can confirm your problem.

**Reply by stefankuehnel** [reply]
Did you fix the problem? If yes, take a look at the following repo. https://github.com/StefanZoneExamples/deta-puppeteer

It contains various security updates and I've installed new (maintained) dependencies. Puppeteer received a major version update on [October 14](https://github.com/puppeteer/puppeteer/releases/tag/v19.0.0). I've created my micro before the release, so that is indeed a problem.

> **Note:** Sadly, with the new update and the new dependencies, you are no longer able to run the project locally.
